File: hackerrank/algorithms/graph-theory/jim-and-his-lan-party/jim-and-his-lan-party.cpp
// Jim and his LAN Party
// https://www.hackerrank.com/challenges/jim-and-his-lan-party
// In C++14.

#include <algorithm>
#include <cassert>
#include <cstdlib>
#include <iostream>
#include <iterator>
#include <memory>
#include <numeric>
#include <unordered_map>
#include <utility>
#include <vector>

namespace {
    void ensure(const bool condition) noexcept
    {
        if (!condition) std::abort();
    }

    constexpr auto not_connected = -1;

    // A network of elements representing users or machiens, each of whom is a
    // member of exactly one group, which can meet/play when all members are
    // connected in a component (possibly shared with members of other groups).
    class Network {
    public:
        // Creates a new network with groups in [0, group_count) and elements
        // in [0, element_groups.size()). Element i is a member of
        // element_groups[i].
        Network(int group_count,
                const std::vector<int>& element_groups) noexcept;

        // Retrieves the times at which each group finished becoming connected.
        // For groups that were never connected, not_connected is returned.
        const std::vector<int>& completion_times() const noexcept;

        // Adds an edge/wire connecting elem1 and elem2.
        void connect(int elem1, int elem2) noexcept;

    private:
        using Contrib = std::unordered_map<int, int>;

        struct ConstructorParametersAreValid { };

        static int validate_parameters_and_return_group_count(
                int group_count,
                const std::vector<int>& element_groups) noexcept;

        static std::unique_ptr<Contrib>
        singleton_contribution(int group) noexcept;

        Network(ConstructorParametersAreValid,
                int group_count,
                const std::vector<int> &element_groups) noexcept;

        void set_initial_group_sizes(
                const std::vector<int>& element_groups) noexcept;

        void set_initial_group_completion_times() noexcept;

        void set_initial_elem_contributions(
                const std::vector<int>& element_groups) noexcept;

        void set_initial_elem_parents() noexcept;

        // Union-find "findset" operation with full path compression.
        int find_set(int elem) noexcept;

        // Makes child a child of parent and merges contributions into parent.
        void join(int parent, int child) noexcept;

        // Merges contributions into whichever contribution map started out
        // larger, recording and removing items that become complete as a
        // result. Returns a handle to the resulting hash, or an empty handle
        // if that has is ultimately empty.
        std::unique_ptr<Contrib>
        merge_contributions(std::unique_ptr<Contrib> contrib1,
                            std::unique_ptr<Contrib> contrib2) noexcept;

        // Merges source into sink. Helper for merge_contributions.
        std::unique_ptr<Contrib>
        do_merge_contributions(std::unique_ptr<Contrib> sink,
                               std::unique_ptr<Contrib> source) noexcept;

        std::vector<int> group_sizes_;
        std::vector<int> group_completion_times_;
        std::vector<std::unique_ptr<Contrib>> elem_contributions_;
        std::vector<int> elem_parents_;
        std::vector<int> elem_ranks_;
        int time_ {0};
    };

    Network::Network(const int group_count,
                     const std::vector<int> &element_groups) noexcept
        : Network{ConstructorParametersAreValid{},
                  validate_parameters_and_return_group_count(group_count,
                                                             element_groups),
                  element_groups}
    {
    }

    const std::vector<int>& Network::completion_times() const noexcept
    {
        return group_completion_times_;
    }

    void Network::connect(int elem1, int elem2) noexcept
    {
        ++time_;

        // Find the ancestors and stop if they are already the same.
        elem1 = find_set(elem1);
        elem2 = find_set(elem2);
        if (elem1 == elem2) return;

        // Union by rank, merging contributions.
        if (elem_ranks_[elem1] < elem_ranks_[elem2]) {
            join(elem2, elem1);
        } else {
            if (elem_ranks_[elem1] == elem_ranks_[elem2]) ++elem_ranks_[elem1];
            join(elem1, elem2);
        }
    }

    int Network::validate_parameters_and_return_group_count(
            const int group_count,
            const std::vector<int>& element_groups) noexcept
    {
        ensure(group_count >= 0);

        ensure(std::all_of(cbegin(element_groups), cend(element_groups),
                           [group_count](const auto group) noexcept {
            return 0 <= group && group < group_count;
        }));

        return group_count;
    }

    auto Network::singleton_contribution(int group) noexcept
            -> std::unique_ptr<Contrib>
    {
        auto contrib = std::make_unique<Contrib>();
        contrib->emplace(group, 1);
        return contrib;
    }

    Network::Network(ConstructorParametersAreValid,
                     const int group_count,
                     const std::vector<int>& element_groups) noexcept
        : group_sizes_(group_count, 0), // Values populated in body.
          group_completion_times_(group_count),
          elem_contributions_(element_groups.size()),
          elem_parents_(element_groups.size()),
          elem_ranks_(element_groups.size(), 0) // Elements start at rank 0.
    {
        set_initial_group_sizes(element_groups);
        set_initial_group_completion_times();
        set_initial_elem_contributions(element_groups);
        set_initial_elem_parents();
    }

    void Network::set_initial_group_sizes(
            const std::vector<int>& element_groups) noexcept
    {
        for (const auto group : element_groups) ++group_sizes_[group];
    }

    void Network::set_initial_group_completion_times() noexcept
    {
        std::transform(cbegin(group_sizes_), cend(group_sizes_),
                begin(group_completion_times_), [&](const auto size) noexcept {
            return size < 2 ? time_ : not_connected;
        });
    }

    void Network::set_initial_elem_contributions(
            const std::vector<int>& element_groups) noexcept
    {
        std::transform(cbegin(element_groups), cend(element_groups),
                begin(elem_contributions_), [&](const auto group) noexcept {
            return group_completion_times_[group] == not_connected
                    ? singleton_contribution(group)
                    : nullptr;
        });
    }

    void Network::set_initial_elem_parents() noexcept
    {
        std::iota(begin(elem_parents_), end(elem_parents_), 0);
    }

    int Network::find_set(int elem) noexcept
    {
        ensure(0 <= elem && elem < elem_parents_.size());

        // Find the ancestor.
        auto leader = elem;
        while (leader != elem_parents_[leader]) leader = elem_parents_[leader];

        // Compress the path.
        while (elem != leader) {
            const auto parent = elem_parents_[elem];
            elem_parents_[elem] = leader;
            elem = parent;
        }

        return leader;
    }

    void Network::join(int parent, int child) noexcept
    {
        elem_parents_[child] = parent;

        elem_contributions_[parent] =
            merge_contributions(std::move(elem_contributions_[parent]),
                                std::move(elem_contributions_[child]));
    }

    auto
    Network::merge_contributions(std::unique_ptr<Contrib> contrib1,
                                 std::unique_ptr<Contrib> contrib2) noexcept
            -> std::unique_ptr<Contrib>
    {
        // If either or both are null, there is nothing to do.
        if (!contrib1) return contrib2;
        if (!contrib2) return contrib1;

        // Merge to the bigger one (so we loop over the smaller).
        return contrib1->size() < contrib2->size()
                ? do_merge_contributions(std::move(contrib2),
                                         std::move(contrib1))
                : do_merge_contributions(std::move(contrib1),
                                         std::move(contrib2));
    }

    auto
    Network::do_merge_contributions(std::unique_ptr<Contrib> sink,
                                    std::unique_ptr<Contrib> source) noexcept
            -> std::unique_ptr<Contrib>
    {
        for (const auto& entry : *source) {
            const auto group = entry.first;
            const auto source_count = entry.second;
            const auto sink_pos = sink->find(group);

            if (sink_pos == sink->end()) {
                sink->emplace(group, source_count);
                continue;
            }

            auto& sink_count = sink_pos->second;

            if (sink_count + source_count < group_sizes_[group]) {
                sink_count = sink_count + source_count;
            } else {
                sink->erase(sink_pos);
                group_completion_times_[group] = time_;
            }
        }

        if (sink->empty()) return nullptr;
        return sink;
    }

    // Reads each player's game. Prepends a 0 for 1-based indexing.
    std::vector<int> read_player_games(const int player_count) noexcept
    {
        ensure(player_count > 0);

        auto element_groups = std::vector<int>(player_count + 1);

        std::copy_n(std::istream_iterator<int>(std::cin), player_count,
                    std::next(begin(element_groups)));

        return element_groups;
    }

    // Skip the extra 0 group and print other groups' completion times.
    void print_game_times(const Network& network) noexcept
    {
        const auto &times = network.completion_times();

        assert(!times.empty());

        std::copy(std::next(cbegin(times)), cend(times),
                  std::ostream_iterator<int>{std::cout, "\n"});
    }
}

int main()
{
    std::ios_base::sync_with_stdio(false);

    // Read the problem parameters.
    auto player_count = -1, game_count = -1, wire_count = -1;
    std::cin >> player_count >> game_count >> wire_count;

    // Check them against problem constraint lower-bounds.
    // (This is more stringent than the UB-avoiding checks Network does.)
    ensure(player_count > 0);
    ensure(game_count > 0);
    ensure(wire_count >= 0);

    // Make the LAN.
    auto network = Network{game_count + 1, // +1 for 1-based indexing
                           read_player_games(player_count)};

    // Apply all the connections.
    while (wire_count-- > 0) {
        auto elem1 = int{}, elem2 = int{};
        std::cin >> elem1 >> elem2;
        network.connect(elem1, elem2);
    }

    print_game_times(network);
}

File: hackerrank/algorithms/graph-theory/jim-and-his-lan-party/jim-and-his-lan-party.cs
// Jim and his LAN Party
// https://www.hackerrank.com/challenges/jim-and-his-lan-party

using System;
using System.Collections.Generic;
using System.Linq;

/// <summary>
/// A network of elements representing users or machines, each of whom is a
/// member of exactly one group, which can meet/play when all members are
/// connected in a component (possibly shared with members of other groups).
/// </summary>
internal sealed class Network {
    /// <summary>
    /// Creates a new network of disconnected elements whose groups are known.
    /// </summary>
    /// <param name="groupCount">Groups are [0, <c>groupCount</c>).</param>
    /// <param name="elementGroups">
    /// Elements are [0, <c>elementGroups.Count</c>).
    /// Element <c>i</c> is a member of <c>elementGroups[i]</c>.
    /// </param>
    internal Network(int groupCount, IList<int> elementGroups)
    {
        _groupSizes = new int[groupCount];
        foreach (var group in elementGroups) ++_groupSizes[group];

        _groupCompletionTimes = _groupSizes
            .Select(size => size < 2 ? _time : NotConnected)
            .ToArray();

        _elemContributions = elementGroups
            .Select(group => _groupCompletionTimes[group] == NotConnected
                                ? SingletonContribution(group)
                                : null)
            .ToArray();

        var totalSize = elementGroups.Count;
        _elemParents = Enumerable.Range(0, totalSize).ToArray();
        _elemRanks = new int[totalSize];
    }

    /// <summary>
    /// The times at which each group finished becoming connected.
    /// For groups that were never connected, <c>NotConnected</c> is given.
    /// </summary>
    internal IEnumerable<int> CompletionTimes
        => _groupCompletionTimes.Select(x => x);

    /// <summary>
    /// Adds an edge/wire connecting <c>elem1</c> and <c>elem2</c>.
    /// </summary>
    internal void Connect(int elem1, int elem2)
    {
        ++_time;

        // Find the ancestors and stop if they are already the same.
        elem1 = FindSet(elem1);
        elem2 = FindSet(elem2);
        if (elem1 == elem2) return;

        // Union by rank, merging contributions.
        if (_elemRanks[elem1] < _elemRanks[elem2]) {
            Join(parent: elem2, child: elem1);
        } else {
            if (_elemRanks[elem1] == _elemRanks[elem2]) ++_elemRanks[elem1];
            Join(parent: elem1, child: elem2);
        }
    }

    internal const int NotConnected = -1;

    private static IDictionary<int, int> SingletonContribution(int group)
        => new Dictionary<int, int> { { group, 1 } };

    /// <summary>
    /// Union-find "findset" operation with full path compression.
    /// </summary>
    private int FindSet(int elem)
    {
        // Find the ancestor.
        var leader = elem;
        while (leader != _elemParents[leader]) leader = _elemParents[leader];

        // Compress the path.
        while (elem != leader) {
            var parent = _elemParents[elem];
            _elemParents[elem] = leader;
            elem = parent;
        }

        return leader;
    }

    /// <summary>Joins sets and merges contributions into the parent.</summary>
    /// <param name="parent">The leader of the new parent set.</param>
    /// <param name="child">The current leader of the new child set.</param>
    private void Join(int parent, int child)
    {
        _elemParents[child] = parent;

        _elemContributions[parent] = MergeContributions(
                                        _elemContributions[parent],
                                        _elemContributions[child]);
        _elemContributions[child] = null;
    }

    /// <summary>
    /// Merges contributions into whichever contribution dictionary started out
    /// larder, recording and removing items that became complete as a result.
    /// </summary>
    /// <param name="contrib1">The first contribution dictionary.</param>
    /// <param name="contrib2">The second contribution dictionary.</param>
    /// <returns>
    /// The resulting map, or <c>null</c> if it is ultimately empty.
    /// </returns>
    /// <remarks>
    /// If both contribution dictionaries have the same number of entries, the
    /// first one, <c>contrib1</c>, is used as the sink.
    /// </remarks>
    private IDictionary<int, int>
    MergeContributions(IDictionary<int, int> contrib1,
                       IDictionary<int, int> contrib2)
    {
        // If either or both is null, there is nothing to do.
        if (contrib1 == null) return contrib2;
        if (contrib2 == null) return contrib1;

        // Merge to the bigger one (so we loop over the smaller).
        return contrib1.Count < contrib2.Count
                ? DoMergeContributions(sink: contrib2, source: contrib1)
                : DoMergeContributions(sink: contrib1, source: contrib2);
    }

    /// <summary>Merges <c>source</c> into <c>sink</c>.</summary>
    /// <remarks>Helper for <see cref="MergeContributions"/>.</remarks>
    private IDictionary<int, int>
    DoMergeContributions(IDictionary<int, int> sink,
                         IDictionary<int, int> source)
    {
        foreach (var sourceEntry in source) {
            var group = sourceEntry.Key;
            var sourceCount = sourceEntry.Value;

            if (!sink.TryGetValue(group, out var sinkCount)) {
                sink.Add(group, sourceCount);
            } else if (sinkCount + sourceCount < _groupSizes[group]) {
                sink[group] = sinkCount + sourceCount;
            } else {
                sink.Remove(group);
                _groupCompletionTimes[group] = _time;
            }
        }

        return sink.Count == 0 ? null : sink;
    }

    private int _time = 0;
    private readonly IList<int> _groupSizes;
    private readonly IList<int> _groupCompletionTimes;
    private readonly IList<IDictionary<int, int>> _elemContributions;
    private readonly IList<int> _elemParents;
    private readonly IList<int> _elemRanks;
}

internal static class Solution {
    private static void Main()
    {
        // Read the problem parameters.
        // (The player count, counts[0], is currently not checked.)
        var counts = ReadRecord().ToArray();
        var gameCount = counts[1];
        var wireCount = counts[2];

        // Make the LAN.
        // The extra 0 element and 0 group faciliate 1-based indexing.
        var network = new Network(gameCount + 1,
                                  ReadRecord().Prepend(0).ToArray());

        // Read and apply all the connections.
        for (var wire = 0; wire < wireCount; ++wire) {
            var elems = ReadRecord().ToArray();
            network.Connect(elems[0], elems[1]);
        }

        // Skip the extra 0 group and print the other gorups' completion times.
        foreach (var time in network.CompletionTimes.Skip(1))
            Console.WriteLine(time);
    }

    /// <summary>Reads a line as a sequence of integers.</summary>
    private static IEnumerable<int> ReadRecord()
        => Console.ReadLine().Split((char[])null,
                                    StringSplitOptions.RemoveEmptyEntries)
                  .Select(int.Parse);
}
